File: problems-originals/1801-1900/1852_distinct-numbers-in-each-subarray/solution.cpp
#include <unordered_map>
#include <vector>

class Solution {
  public:
    // One frequency map slides with the window; the running count of values
    // whose frequency is nonzero is the answer per window.
    std::vector<int> distinctNumbers(std::vector<int> &nums, int k) {
        int n = nums.size();
        std::vector<int> ans(n - k + 1);
        std::unordered_map<int, int> freq;
        freq.reserve(n * 2);
        int distinct = 0;
        for (int i = 0; i < n; i++) {
            distinct += freq[nums[i]]++ == 0;
            if (i >= k && --freq[nums[i - k]] == 0) {
                distinct--;
            }
            if (i >= k - 1) {
                ans[i - k + 1] = distinct;
            }
        }
        return ans;
    }
};

File: problems-originals/1801-1900/1852_distinct-numbers-in-each-subarray/solution.rs
use std::collections::HashMap;

impl Solution {
    // One frequency map slides with the window; the running count of values
    // whose frequency is nonzero is the answer per window.
    pub fn distinct_numbers(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let k = k as usize;
        let mut ans = Vec::with_capacity(nums.len() - k + 1);
        let mut freq: HashMap<i32, i32> = HashMap::new();
        let mut distinct = 0i32;
        for (i, &v) in nums.iter().enumerate() {
            let e = freq.entry(v).or_insert(0);
            *e += 1;
            if *e == 1 {
                distinct += 1;
            }
            if i >= k {
                let left = nums[i - k];
                let e = freq.get_mut(&left).unwrap();
                *e -= 1;
                if *e == 0 {
                    distinct -= 1;
                }
            }
            if i + 1 >= k {
                ans.push(distinct);
            }
        }
        ans
    }
}
